### PR TITLE
Refactor command handlers

### PR DESCRIPTION
**Pull request by bakurits**
#### Summary

- Split ExecuteCommand into separate command handlers.
- Made map of [command][handler]
- Resolved some unhandled error warnings

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-plugin-github/issues/211

### CODE REVIEW

**Comment by bakurits** [comment]
I think additional tests are needed. I don't know much about commands behavior, so this should be separate issued

File: server/command.go [hunk 1]
```diff
@@ -95,10 +95,205 @@ func (p *Plugin) postCommandResponse(args *model.CommandArgs, text string) {
...
+	text := "###### Mattermost GitHub Plugin - Slash Command Help\n" + strings.Replace(COMMAND_HELP, "|", "`", -1)
+	p.postCommandResponse(args, text)
+	return &model.CommandResponse{}, nil
+}
+func handleEmpty(p *Plugin, _ *plugin.Context, args *model.CommandArgs, _ []string, _ *github.Client, _ context.Context, _ *GitHubUserInfo) (*model.CommandResponse, *model.AppError) {
+	text := "###### Mattermost GitHub Plugin - Slash Command Help\n" + strings.Replace(COMMAND_HELP, "|", "`", -1)
+	p.postCommandResponse(args, text)
+	return &model.CommandResponse{}, nil
+}
+func handleSettings(p *Plugin, _ *plugin.Context, args *model.CommandArgs, parameters []string, _ *github.Client, _ context.Context, userInfo *GitHubUserInfo) (*model.CommandResponse, *model.AppError) {
```
Comment (hanzei) [inline server/command.go]: Why is this a function and not a method of `*Plugin`?

File: server/command.go [hunk 2]
```diff
@@ -95,10 +95,205 @@ func (p *Plugin) postCommandResponse(args *model.CommandArgs, text string) {
...
+	err := p.storeGitHubUserInfo(userInfo)
+	if err != nil {
+		mlog.Error(err.Error())
+	}
+
+	p.postCommandResponse(args, "Settings updated.")
+	return &model.CommandResponse{}, nil
+}
+
+type CommandHandleFunc func(p *Plugin, c *plugin.Context, args *model.CommandArgs, parameters []string, githubClient *github.Client, ctx context.Context, userInfo *GitHubUserInfo) (*model.CommandResponse, *model.AppError)
```
Comment (hanzei) [inline server/command.go]: Why is the same `ctx` for the commands needed? Can they use there own?

Comment (hanzei) [inline server/command.go]: Is it worth to not pass `*github.Client` every time and instead let he handlers create it themself?

Comment (hanzei) [inline server/command.go]: Every handler method returns `&model.CommandResponse{}, nil`. Can we unity the code?

File: server/command.go [hunk 3]
```diff
@@ -95,10 +95,205 @@ func (p *Plugin) postCommandResponse(args *model.CommandArgs, text string) {
...
+	} else if setting == SETTING_REMINDERS {
+		userInfo.Settings.DailyReminder = value
+	}
+
+	err := p.storeGitHubUserInfo(userInfo)
+	if err != nil {
+		mlog.Error(err.Error())
+	}
+
+	p.postCommandResponse(args, "Settings updated.")
```
Comment (hanzei) [inline server/command.go]: Most methods call `p.postCommandResponse` to send a method to the user. Could this message string make sense as a return value for the handlers?

File: server/command.go (hunk 1 above)
Comment (bakurits) [reply server/command.go]: In that case, `commandHandlers` map should be inside `Plugin` and initialize with it. Something like that
```go
p.commandHandlers = map[string]CommandHandleFunc{
		"subscribe":   p.handleSubscribe,
		"unsubscribe": p.handleUnSubscribe,
		"disconnect": p.handleDisconnect,
		"todo":        p.handleTodo,
		"me":          p.handleMe,
		"help":       p.handleHelp,
		"":            p.handleEmpty,
		"settings":    p.handleSettings,
	}
```

File: server/command.go (hunk 2 above)
Comment (bakurits) [reply server/command.go]: This code should be a separate function then.
```go
        var githubClient *github.Client

	info, apiErr := p.getGitHubUserInfo(args.UserId)
	if apiErr != nil {
		text := "Unknown error."
		if apiErr.ID == API_ERROR_ID_NOT_CONNECTED {
			text = "You must connect your account to GitHub first. Either click on the GitHub logo in the bottom left of the screen or enter `/github connect`."
		}
		p.postCommandResponse(args, text)
		return &model.CommandResponse{}, nil
	}

	githubClient = p.githubConnect(*info.Token)
```

File: server/command.go (hunk 3 above)
Comment (bakurits) [reply server/command.go]: so what you are saying is we should return message from handlers and no more
```go
p.postCommandResponse(args, message)```
lines inside handlers, we will do that after returning message

**Comment by hanzei** [comment]
@bakurits FYI: The last commit caused a build failure:
```
./command.go:162:17: assignment mismatch: 3 variables but parseOwnerAndRepo returns 2 values
make: *** [Makefile:71: server] Error 2
```

**Comment by bakurits** [comment]
> @bakurits FYI: The last commit caused a build failure:
> 
> ```
> ./command.go:162:17: assignment mismatch: 3 variables but parseOwnerAndRepo returns 2 values
> make: *** [Makefile:71: server] Error 2
> ```

wow I think my utils.go file is out of date parseOwnerAndRepo is returning 3 variables in my version

File: server/command.go [hunk 4]
```diff
@@ -135,174 +305,12 @@ func (p *Plugin) ExecuteCommand(c *plugin.Context, args *model.CommandArgs) (*mo
...
-		} else if setting == SETTING_REMINDERS {
-			info.Settings.DailyReminder = value
-		}
-
-		p.storeGitHubUserInfo(info)
-
-		p.postCommandResponse(args, "Settings updated.")
-		return &model.CommandResponse{}, nil
+	if f, ok := p.CommandHandlers[command]; ok {
+		message, err := f(c, args, parameters, info)
```
Comment (hanzei) [inline server/command.go]: Is there any place where `err != nil`? If not, can we drop the error return value.

Comment (bakurits) [reply server/command.go]: `err` is always nil. I can remove it

**Comment by mickmister** [comment]
@DHaussermann This will need a regression test for each slash command/subcommand

**Comment by DHaussermann** [comment]
@bakurits When I build and deploy this I'm seeing that the commands that were moved no longer work. No changes where made to how the commands must be used correct?

For example `/github help`  is no longer recognized. 

GitHub-Command

Can you see if you can repro this? As far as I can tell I have the latest commit so, I'm not sure what's happening.

**Comment by bakurits** [comment]
@DHaussermann I think I've found the bug. Checks don't work, don't know why.

I think we should add unit tests for command

File: server/command.go [hunk 5]
```diff
@@ -96,10 +96,181 @@ func (p *Plugin) postCommandResponse(args *model.CommandArgs, text string) {
...
+		return fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+	}
+
+	if err := p.Subscribe(ctx, githubClient, args.UserId, owner, repo, args.ChannelId, features, flags); err != nil {
+		return err.Error()
+	}
+
+	return fmt.Sprintf("Successfully subscribed to %s.", repo)
+}
+func (p *Plugin) handleUnSubscribe(_ *plugin.Context, args *model.CommandArgs, parameters []string, _ *GitHubUserInfo) string {
```
Comment (larkox) [inline server/command.go:176]: nit: handleUnSubscribe -> handleUnsubscribe

**Comment by hanzei** [comment]
Great work @bakurits :tada:

**Pull request merged**